MessageControl/MessageControl.h

```diff
@@ -248,30 +306,65 @@ namespace Plugin {
 
             _outputLock.Unlock();
         }
-        void Message(const Core::Messaging::Metadata::type type,
-            const string & module, const string& category, const string & fileName,
-            const uint16_t lineNumber, const string & className,
-            const uint64_t timestamp, const string & message) {
 
+        void Message(const Core::Messaging::Metadata& metadata, const string& message)
+        {
             // Time to start sending it to all interested parties...
             _outputLock.Lock();
 
             for (auto& entry : _outputDirector) {
-                entry->Message(type, module, category, fileName, lineNumber, className, timestamp, message);
+                entry->Message(metadata, message);
             }
 
-            _webSocketExporter.Message(type, module, category, fileName, lineNumber, className, timestamp, message);
+            _webSocketExporter.Message(metadata, message);
 
             _outputLock.Unlock();
         }
-        void Attach(const uint32_t id) {
+        
+    public:
+        uint32_t Callback(Plugin::MessageControl::ICollect::ICallback* callback)
+        {
+            _adminLock.Lock();
+
+            ASSERT((_callback != nullptr) ^ (callback != nullptr));
+
+            if (_callback != nullptr) {
+                _worker.Block();
+                _client.SkipWaiting();
+                _adminLock.Unlock();
+                _worker.Wait(Core::Thread::BLOCKED, Core::infinite);
+                _adminLock.Lock();
+            }
+
+            _callback = callback;
+
+            if (_callback != nullptr) {
+                _worker.Run();
+            }
+
+            _adminLock.Unlock();
+
+            return (Core::ERROR_NONE);
+        }
+
+        void Attach(const uint32_t id)
+        {
             _adminLock.Lock();
-            _collect->Attach(id);
+
+            _client.AddInstance(id);
+            Cleanup();
+
             _adminLock.Unlock();
         }
-        void Detach(const uint32_t id) {
+        
+        void Detach(const uint32_t id)
+        {
             _adminLock.Lock();
-            _collect->Detach(id);
+
+            _client.RemoveInstance(id);
+            _cleaning.emplace_back(id);
+            Cleanup();
+            
             _adminLock.Unlock();
 
             if (id == _connectionId) {
```

Comment:
Can we get rid of the `_connectionId` class member entirely then, if it is not need for `IRemoteConnection`, or does it need to stay for this part?